src/icp_v2.cc

```diff
@@ -670,38 +649,37 @@ icpHandleUdp(int sock, void *)
             if (xerrno != ECONNREFUSED && xerrno != EHOSTUNREACH)
 #endif
                 debugs(50, DBG_IMPORTANT, "icpHandleUdp: FD " << sock << " recvfrom: " << xstrerr(xerrno));
-
             break;
         }
 
         ++(*N);
         icpCount(buf, RECV, (size_t) len, 0);
         buf[len] = '\0';
-        debugs(12, 4, "icpHandleUdp: FD " << sock << ": received " <<
-               (unsigned long int)len << " bytes from " << from);
-
-#ifdef ICP_PACKET_DUMP
 
-        icpPktDump(buf);
-#endif
-
-        if ((size_t) len < sizeof(icp_common_t)) {
-            debugs(12, 4, "icpHandleUdp: Ignoring too-small UDP packet");
-            break;
-        }
+        icp_common_t pkt(buf, len);
 
-        icp_version = (int) buf[1]; /* cheat! */
+        debugs(12, 2, "ICP Client remote=" << from << " FD " << sock);
+        debugs(12, 2, "ICP Client REQUEST:\n---------\n" <<
+               icp_opcode_str[pkt.opcode] << " #" << pkt.reqnum << " ICP/" << pkt.version << std::endl <<
+               ":length: " << pkt.length << std::endl <<
+               ":flags: " << asHex(pkt.flags) <<
+               Debug::Extra << Raw("payload", (buf+sizeof(icp_common_t)), pkt.length).minLevel(DBG_DATA).hex() <<
+               "\n----------");
 
         if (icpOutgoingConn->local == from)
             // ignore ICP packets which loop back (multicast usually)
-            debugs(12, 4, "icpHandleUdp: Ignoring UDP packet sent by myself");
-        else if (icp_version == ICP_VERSION_2)
+            throw TextException("UDP packet sent by myself", Here());
+        else if (pkt.version == ICP_VERSION_2)
             icpHandleIcpV2(sock, from, buf, len);
-        else if (icp_version == ICP_VERSION_3)
+        else if (pkt.version == ICP_VERSION_3)
             icpHandleIcpV3(sock, from, buf, len);
         else
-            debugs(12, DBG_IMPORTANT, "WARNING: Unused ICP version " << icp_version <<
-                   " received from " << from);
+            throw TextException(ToSBuf("unknown version ", pkt.version), Here());
+
+        } catch (...) {
+            debugs(12, 2, "WARNING: ICP remote=" << from << " FD " << sock << ": " << CurrentException << " (ignoring)");
```

Comment:
As discussed [earlier](https://github.com/squid-cache/squid/pull/827#discussion_r638947024), this hides the "unknown ICP version" error messages from the admin. That message was probably a good canary for an attack on the ICP port with random data.

> Amos: The high level message about unknown ICP version being ignored does not need to be logged at level-1.

If the PR hides that error, I am not going to block it on those grounds because I do not have the time to argue and detail a good solution for the overall problem. However, if the PR hides that error, its description must disclose and justify that important reporting change. This is the [second](https://github.com/squid-cache/squid/pull/827#discussion_r638947024) time I am requesting that disclosure.

As always, to save time, please add your justifications to the PR description rather than (just to) this change request thread.